clippy_lints/src/cognitive_complexity.rs

```diff
@@ -48,190 +52,258 @@ impl LintPass for CognitiveComplexity {
+                                }
+                            }
+                        },
+                        ExprKind::Paren(expr) | ExprKind::Unary(_, expr) => leftmost_lop(&expr),
+                        _ => None,
+                    }
+                }
+
+                match binop.node {
+                    BinOpKind::And => {
+                        if let Some(LogOp::Or) = leftmost_lop(r_ex) {
+                            self.score_ni_structure();
+                        }
+                        if let Some(LogOp::Or) = rightmost_lop(l_ex) {
+                            self.score_ni_structure();
                         }
                     },
-                    _ => (),
+                    BinOpKind::Or => {
+                        if let Some(LogOp::And) = leftmost_lop(r_ex) {
+                            self.score_ni_structure();
+                        }
+                        if let Some(LogOp::And) = rightmost_lop(l_ex) {
+                            self.score_ni_structure();
+                        }
+                    },
+                    _ => {},
                 }
+
+                walk_expr(self, ex);
             },
-            ExprKind::Closure(.., _) => (),
-            ExprKind::Binary(op, _, _) => {
-                walk_expr(self, e);
-                match op.node {
-                    BinOpKind::And | BinOpKind::Or => self.short_circuits += 1,
-                    _ => (),
-                }
+
+            // ### Probably non-scoring cases ###
+            // FIXME: make this explicit in the Draft.
+
+            // THOUGHT: these shouldn't increase the score.
+            // ExprKind::Path(..) => {},
+            // ExprKind::Range(..) => {},
+            // ExprKind::Index(..) => {},
+            // ExprKind::Field(..) => {},
+            // ExprKind::Tup(..) => {},
+            // ExprKind::Box(..) => {},
+            // ExprKind::ObsoleteInPlace(..) => {},
+            // ExprKind::Array(..) => {},
+            // ExprKind::AssignOp(..) => {},
+            // ExprKind::Assign(..) => {},
+            // ExprKind::Lit(..) => {},
+            // ExprKind::Struct(..) => {},
+            // ExprKind::Repeat(..) => {},
+            // ExprKind::AddrOf(..) => {},
+
+            // FIXME: examine these more closely
+            // ExprKind::Paren(..) => {},
+            // ExprKind::Unary(..) => {},
+            // ExprKind::Ret(..) => {},
+
+            // ### Undiscussed Cases ###
+            // FIXME: discuss and solve each one of these.
+            // ExprKind::Cast(..) => {},
+            // ExprKind::Type(..) => {},
+            // ExprKind::WhileLet(..) => {},
+            // ExprKind::Async(..) => {},
+            // ExprKind::Try(..) => {},
+            // ExprKind::TryBlock(..) => {},
+            // ExprKind::Yield(..) => {},
+            // ExprKind::InlineAsm(..) => {},
+            // ExprKind::Err => {},
+            _ => {
+                // walk_expr(self, ex);
```

Comment:
You need to walk expressions. Visitors don't recurse automatically like lintpasses do.

Basically you're treating 

```rust
fn foo() -> i32 {
    { /* complex code */ } as u32
}
```

as having a CoC of 1, because you're not looking at casts

Reply:
Ahh, yes! I forgot this one. Thank you 😊

Reply:
Uncommented ^^